src/s3/async_put_object.rs

```diff
@@ -107,24 +107,17 @@ impl<'a> AsyncWrite for AsyncPutObject<'a> {
                     .acl(ObjectCannedAcl::BucketOwnerFullControl)
                     .send();
                 self.state = PutObjectState::Closing(Box::pin(fut));
+                // Manually trigger a wake so that the executor immediately
+                // polls us again, which will take us into the `Closing` block.
                 cx.waker().wake_by_ref();
                 Poll::Pending
             }
             PutObjectState::Closing(ref mut fut) => {
-                match Pin::new(fut).poll(cx) {
-                    Poll::Ready(x) => {
-                        self.state = PutObjectState::Closed;
-                        match x {
-                            Ok(_) => Poll::Ready(Ok(())),
-                            Err(e) => Poll::Ready(Err(Error::new(ErrorKind::Other, e))),
-                        }
-                    }
-                    Poll::Pending => {
-                        // Tell the async executor that we're ready to try again whenever it is!
-                        cx.waker().wake_by_ref();
-                        Poll::Pending
-                    }
-                }
+                let result = ready!(Pin::new(fut).poll(cx))
+                    .map(|_| ())
+                    .map_err(|e| Error::new(ErrorKind::Other, e));
```

Comment:
FWIW I think using `Result` helper methods here makes the overall logic a bit more legible than doing a `match`, but your mileage may vary.

Reply:
Yep, that works for me 👍